mlir/lib/Catalyst/Transforms/TBAAPatterns.cpp

```diff
@@ -0,0 +1,129 @@
+                                             adaptor.getIndices(), rewriter);
+        auto op = rewriter.replaceOpWithNewOp<LLVM::LoadOp>(
+            loadOp, typeConverter->convertType(type.getElementType()), dataPtr, 0, false,
+            loadOp.getNontemporal());
+        auto baseType = loadOp.getMemRefType().getElementType();
+
+        mlir::LLVM::TBAATagAttr tag;
+
+        if (isa<IndexType>(baseType) || dyn_cast<IntegerType>(baseType)) {
+            tag = tree->getTag("int");
+            op.setTBAATags(ArrayAttr::get(loadOp.getContext(), tag));
+        }
+        else if (dyn_cast<FloatType>(baseType)) {
+            if (baseType.isF32()) {
+                tag = tree->getTag("float");
+            }
+            else if (baseType.isF64()) {
+                tag = tree->getTag("double");
+            }
+
+            op.setTBAATags(ArrayAttr::get(loadOp.getContext(), tag));
+        }
+        else if (dyn_cast<MemRefType>(baseType)) {
+            tag = tree->getTag("any pointer");
+
+            op.setTBAATags(ArrayAttr::get(loadOp.getContext(), tag));
+        }
+        return success();
+    }
+
+  private:
+    catalyst::TBAATree *tree = nullptr;
+};
+
+struct MemrefStoreTBAARewritePattern : public ConvertOpToLLVMPattern<memref::StoreOp> {
+    using ConvertOpToLLVMPattern<memref::StoreOp>::ConvertOpToLLVMPattern;
+
+    template <typename... Args>
+    MemrefStoreTBAARewritePattern(catalyst::TBAATree &tree, Args &&...args)
+        : ConvertOpToLLVMPattern(std::forward<Args>(args)...), tree(&tree){};
+
+    LogicalResult matchAndRewrite(memref::StoreOp storeOp, memref::StoreOpAdaptor adaptor,
+                                  ConversionPatternRewriter &rewriter) const override
+    {
+        auto type = storeOp.getMemRefType();
+        auto baseType = type.getElementType();
+
+        Value dataPtr = getStridedElementPtr(storeOp.getLoc(), type, adaptor.getMemref(),
+                                             adaptor.getIndices(), rewriter);
+        auto op = rewriter.replaceOpWithNewOp<LLVM::StoreOp>(storeOp, adaptor.getValue(), dataPtr,
+                                                             0, false, storeOp.getNontemporal());
+
+        mlir::LLVM::TBAATagAttr tag;
+
+        if (isa<IndexType>(baseType) || dyn_cast<IntegerType>(baseType)) {
+            tag = tree->getTag("int");
+            op.setTBAATags(ArrayAttr::get(storeOp.getContext(), tag));
+        }
+        else if (dyn_cast<FloatType>(baseType)) {
+            if (baseType.isF32()) {
+                tag = tree->getTag("float");
+            }
+            else if (baseType.isF64()) {
+                tag = tree->getTag("double");
+            }
+
+            op.setTBAATags(ArrayAttr::get(storeOp.getContext(), tag));
+        }
+        else if (dyn_cast<MemRefType>(baseType)) {
+            tag = tree->getTag("any pointer");
+
+            op.setTBAATags(ArrayAttr::get(storeOp.getContext(), tag));
+        }
```

Comment:
Consider adding a check here that an else is unreachable or something similar.

Reply:
I add unreachable as well :+1: